### PR TITLE
Share Cache Service logs between multiple concurrent downloads

### PR DESCRIPTION
**Pull request by kraih**
This fixes the problem where only the first job requesting an asset from the Cache Service gets the logs. Every other job requesting the same asset would have gotten `Asset "..." was downloaded by another job, details are therefore unavailable here` previously. Making it non-trivial for test developers to find the cause for many cache service problems.

There was also a race condition in `25-cache-service.t`, so that test should be much more stable now.

### CODE REVIEW

**Comment by kraih** [comment]
Since this patch adds a new migration to the cache service (for the very first time) it will need testing on the staging machines before it can be merged.

**Comment by kraih** [comment]
Btw. i've included the `$a`/`$b` name fix in this PR because it is very harmless and the cache service tests were actually the worst offenders.

File: lib/OpenQA/CacheService.pm [hunk 1]
```diff
@@ -140,3 +141,30 @@ Enqueue the task. It acceps a POST JSON payload of the form:
...
+    `id` INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+    `lock` TEXT,
+    `job_id` TEXT,
+    `created` DATETIME NOT NULL DEFAULT current_timestamp
+);
+CREATE INDEX IF NOT EXISTS downloads_lock on downloads (lock);
+CREATE INDEX IF NOT EXISTS downloads_created on downloads (created);
+
+-- 2 down
+DROP TABLE downloads;
```
Comment (okurz) [inline lib/OpenQA/CacheService.pm]: curious, how did you come up with this `_DATA_` section?

Comment (kraih) [reply lib/OpenQA/CacheService.pm]: What do you mean? It was previously in `OpenQA::CacheService::Model::Cache` for historic reasons, but now the schema spans two classes (`OpenQA::CacheService::Model::Downloads`). So moving it into the application class made sense.

Comment (kraih) [reply lib/OpenQA/CacheService.pm]: I could also have moved the migrations into an external SQL file, similar to how [Minion](https://github.com/mojolicious/minion/blob/master/lib/Minion/Backend/resources/migrations/pg.sql) does it, but we have no naming convention for that in openQA yet and i wasn't too keen on arguing about it. 😉

Comment (kraih) [reply lib/OpenQA/CacheService.pm]: If you're not seen this migration format yet, it's from [Mojo::SQLite::Migrations](https://metacpan.org/pod/Mojo::SQLite::Migrations#DESCRIPTION). Really simple, just one SQL file with comments separating the versions.

**Comment by kraih** [comment]
I am a little worried about adding more complexity to the Cache Service model after spending so much time on simplifying it with the last refactoring. But having the full download logs in `autoinst-log.txt` for every job is probably worth it.

**Comment by kraih** [comment]
And @codecov lies... i did not touch ObsRsync! This is 100% an increase in test coverage.

**Comment by kraih** [comment]
@Martchus There's [new tests](https://github.com/os-autoinst/openQA/pull/2615/files#diff-12e1fddde2e0c415e4a419740c7f9931R522) for both cases, assets and rsync. Change is so large it's GitHub has it hidden by default. 😆

**Comment by Martchus** [comment]
I just see checks for `qr/Asset "sle.+" was downloaded by #\d+, details are therefore unavailable here/`. But isn't that the message we want to avoid and instead include the defaults? Or is this really just about including a reference? Is the reference a Minion, GRU or openQA job id? It would make sense to specify that in the message to avoid confusion.

**Comment by kraih** [comment]
@Martchu There is two checks, the first does indeed check for the `...details are therefore unavailable...` in the concurrent job meta data. Because that's our fallback in case the read download job ever vanishes before the concurrent download job. But the next check after that ensures the status API actually sends the [real download status](https://github.com/os-autoinst/openQA/pull/2615/files#diff-12e1fddde2e0c415e4a419740c7f9931R560).

File: lib/OpenQA/CacheService.pm (hunk 1 above)
Comment (okurz) [reply lib/OpenQA/CacheService.pm]: ok, thx for explanation. As I stated, I was just "curious", not at all critic or something.

Comment (kraih) [reply lib/OpenQA/CacheService.pm]: The question was just a bit vague.

**Comment by kraih** [comment]
@okurz Sure, here are 3 jobs that demonstrate the cases.


http://openqa-staging-1.qa.suse.de/tests/342/file/autoinst-log.txt
http://openqa-staging-1.qa.suse.de/tests/343/file/autoinst-log.txt
http://openqa-staging-1.qa.suse.de/tests/344/file/autoinst-log.txt

At first glance you will see that they mostly share the same log messages regarding the Cache Service. But the job number are slightly different.
```
[2019-12-19T11:41:24.0921 CET] [info] [pid:20790] Downloading SLES-15-SP1-x86_64-Installtest.qcow2, request #10183 sent to Cache Service
[2019-12-19T11:57:15.0145 CET] [info] [pid:20790] Download of SLES-15-SP1-x86_64-Installtest.qcow2 processed:
[info] [#10181] Cache size of "/var/lib/openqa/cache" is 6.3GiB, with limit 20GiB
[info] [#10181] Downloading "SLES-15-SP1-x86_64-Installtest.qcow2" from "openqa-staging-1.qa.suse.de/tests/342/asset/hdd/SLES-15-SP1-x86_64-Installtest.qcow2"
[info] [#10181] Size of "/var/lib/openqa/cache/openqa-staging-1.qa.suse.de/SLES-15-SP1-x86_64-Installtest.qcow2" is 1GiB, with ETag ""4273cc00-597ebb5a2b000""
[info] [#10181] Download of "/var/lib/openqa/cache/openqa-staging-1.qa.suse.de/SLES-15-SP1-x86_64-Installtest.qcow2" successful, new cache size is 7.4GiB

[2019-12-19T11:57:15.0168 CET] [info] [pid:20790] Downloading SLE-15-SP1-Installer-DVD-x86_64-GM-DVD1.iso, request #10184 sent to Cache Service
[2019-12-19T11:57:20.0242 CET] [info] [pid:20790] Download of SLE-15-SP1-Installer-DVD-x86_64-GM-DVD1.iso processed:
[info] [#10184] Cache size of "/var/lib/openqa/cache" is 7.4GiB, with limit 20GiB
[info] [#10184] Downloading "SLE-15-SP1-Installer-DVD-x86_64-GM-DVD1.iso" from "openqa-staging-1.qa.suse.de/tests/344/asset/iso/SLE-15-SP1-Installer-DVD-x86_64-GM-DVD1.iso"
[info] [#10184] Content of "/var/lib/openqa/cache/openqa-staging-1.qa.suse.de/SLE-15-SP1-Installer-DVD-x86_64-GM-DVD1.iso" has not changed, updating last use
```
Here for example we have the log output of openQA job 344. You can see that it enqueued the cache service job 10183 to download `SLES-15-SP1-x86_64-Installtest.qcow2`. But a download of that asset was already in progress (requested by openQA job 342), so you see the cache service job id 10181 in the logs. Here the download logs from the other job have been shared.

The same can be seen for the rsync later on, but there it was actually openQA job 344 that requested it first with the cache service job 10187. And the other two jobs have shared those logs.
```
[2019-12-19T11:57:24.0354 CET] [info] [pid:20749] Rsync from 'rsync://openqa-staging-1.qa.suse.de/tests' to '/var/lib/openqa/cache/openqa-staging-1.qa.suse.de', request #10189 sent to Cache Service
[2019-12-19T11:57:29.0435 CET] [info] [pid:20749] Output of rsync:
[info] [#10187] Calling: rsync -avHP rsync://openqa-staging-1.qa.suse.de/tests/ --delete /var/lib/openqa/cache/openqa-staging-1.qa.suse.de/tests/
receiving incremental file list
sle/.git/
sle/.git/FETCH_HEAD

              0   0%    0.00kB/s    0:00:00  
            118 100%  115.23kB/s    0:00:00 (xfr#1, ir-chk=1028/1059)
sle/.git/logs/refs/remotes/origin/master
...
```

And finally there is another asset download where the asset was already cached and the requests were handled so fast that the logs did not need to be shared at all. So each openQA job got their very own logs for `SLE-15-SP1-Installer-DVD-x86_64-GM-DVD1.iso`.
```
[2019-12-19T11:57:15.0168 CET] [info] [pid:20790] Downloading SLE-15-SP1-Installer-DVD-x86_64-GM-DVD1.iso, request #10184 sent to Cache Service
[2019-12-19T11:57:20.0242 CET] [info] [pid:20790] Download of SLE-15-SP1-Installer-DVD-x86_64-GM-DVD1.iso processed:
[info] [#10184] Cache size of "/var/lib/openqa/cache" is 7.4GiB, with limit 20GiB
[info] [#10184] Downloading "SLE-15-SP1-Installer-DVD-x86_64-GM-DVD1.iso" from "openqa-staging-1.qa.suse.de/tests/344/asset/iso/SLE-15-SP1-Installer-DVD-x86_64-GM-DVD1.iso"
[info] [#10184] Content of "/var/lib/openqa/cache/openqa-staging-1.qa.suse.de/SLE-15-SP1-Installer-DVD-x86_64-GM-DVD1.iso" has not changed, updating last use
```

Previously two of those three jobs would have just shown an `Asset "..." was downloaded by another job, details are therefore unavailable here` message without any details about the downloads.

**Comment by kraih** [comment]
I did not encounter any problems on the staging machines.

**Comment by okurz** [comment]
also tried it out locally with a non-caching worker, looks good

**Pull request merged**